internal/controller/apiextensions/composite/composed.go

```diff
@@ -178,28 +179,49 @@ func (cdf *APIConnectionDetailsFetcher) FetchConnectionDetails(ctx context.Conte
 	}
 
 	for _, d := range t.ConnectionDetails {
-		if d.Name != nil && d.Value != nil {
-			conn[*d.Name] = []byte(*d.Value)
-			continue
-		}
+		switch {
```

Comment:
Given we now have a few mutually exclusive secret sources (e.g. value, from connection secret key, and from field path) I think it's probably time to add a [discriminator field](https://github.com/kubernetes/kubernetes/issues/35345) here, similar to https://github.com/crossplane/crossplane/issues/1974. This would let us switch on that field rather than trying to infer intent from which values were and were not set.

I'd suggest something like:

```yaml
resources:
- base:
    # YAML..
  connectionDetails:
  - type: FromConnectionSecretKey  # The default, since this field needs to be optional
    fromConnectionSecretKey: somekey
  - type: FromFieldPath
    fromFieldPath: "status.interestingField"
```

I wonder if we should use something more specificly named than `FromFieldPath` to make it clear that the path is within the composed resource, not the composite. We've previously shied away from `FromComposedFieldPath` because it's easy for folks to confuse "composed" and "composite" at a glance.

Reply:
How about `FromResourceFieldPath` or `ValueFromResourceFieldPath`? I think it should also note that `name` must be specified when set.

Reply:
Sounds good to me, I'll make the relevant changes shortly :)